### PR TITLE
KAFKA-17064: New consumer assign should update assignment in background thread

### PR DESCRIPTION
**Pull request by FrankYang0529**
With the new async consumer, the subscriptionState object is shared between the app thread and the background thread, but in principle all updates to the assignment should happen in the background thread, to avoid race conditions.

We should consider moving the assignment update to the background thread, as part of the `AssignmentChangeEvent`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by FrankYang0529** [comment]
Hi @lianetm, I have two questions about this enhancmnet.

1. Do we want to remove `NewTopicsMetadataUpdateRequestEvent`? It's not used after this PR.
2. Do we want to remove `offsets` in `AssignmentChangeEvent`? `ApplicationEventProcessor#process` doesn't use this data. It gets the data from `SubscriptionState#allConsumed` directly in `ApplicationEventProcessor#maybeAutoCommitAsync`.

**Comment by lianetm** [comment]
Hey @FrankYang0529. Thanks for the patch! I haven't reviewed this PR yet, but from a quick look: 
1. yes, let's remove the unused event. Makes sense that we don't need it anymore since all happens in the background now.
2. yes, seems sensible to remove the unused `offsets` and get the offsets to commit from the subscriptionState#allConsumed directly (that we have handy when processing the assignment change event)

I'll take a closer look at this PR asap in the next days. Thanks!

**Comment by FrankYang0529** [comment]
Hi @lianetm, thanks for the answer. I will covert the PR to ready for review again after I remove unused class and variable. Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -1443,12 +1443,7 @@ public void assign(Collection<TopicPartition> partitions) {
             // be no following rebalance.
             //
             // See the ApplicationEventProcessor.process() method that handles this event for more detail.
-            applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
-
-            log.info("Assigned to partition(s): {}", partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-            if (subscriptions.assignFromUser(new HashSet<>(partitions)))
-                applicationEventHandler.add(new NewTopicsMetadataUpdateRequestEvent());
+            applicationEventHandler.add(new AssignmentChangeEvent(time.milliseconds(), partitions));
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Adding this only means that we will potentially return from the `assign()` without having really updated the subscription state. I wonder if we should go with the same approach we have on the unsubscribe [here](https://github.com/apache/kafka/blob/901c656a1da16c8473200c93e57991004fc1371d/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1487-L1492), that needs to deal with a very similar situation (and actually, assign with empty triggers unsubscribe and goes down that path that seems more in-line with what the legacy does). 

The trick would be to add the assign event, but then poll for background events until the assign event completes (so we know that the subscription state was updated). Makes sense?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java [hunk 2]
```diff
@@ -16,33 +16,31 @@
...
 package org.apache.kafka.clients.consumer.internals.events;
 
-import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.common.TopicPartition;
 
-import java.util.Collections;
-import java.util.Map;
+import java.util.Collection;
 
 public class AssignmentChangeEvent extends ApplicationEvent {
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java]: To be able to know when the assignment change takes effect, we should probably have to make this extend CompletableApplicationEvent (Void probably? we just need to know that it completed)

File: clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java [hunk 3]
```diff
@@ -453,7 +453,8 @@ public void testSubscription(GroupProtocol groupProtocol) {
 
         consumer.assign(singletonList(tp0));
         assertTrue(consumer.subscription().isEmpty());
-        assertEquals(singleton(tp0), consumer.assignment());
+        TestUtils.waitForCondition(() -> consumer.assignment().equals(singleton(tp0)),
+                "Expected consumer to be assigned the partition");
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java]: If the suggestion on my [comment](https://github.com/apache/kafka/pull/16673/files#r1697100482) above works, I would expect we won't need to change this test like this (shows a rather undesired behaviour of the assign I would say). We would end with the test unchanged probably, with the assignment effectively changed when assign() returns for both consumers.

**Comment by lianetm** [comment]
Hey @FrankYang0529 , thanks for the changes! Left some high level comments for consideration, let me know what you think. Thanks!

**Comment by FrankYang0529** [comment]
Hi @lianetm, thanks for the suggestion! I implement `CompletableApplicationEvent` for `AssignmentChangeEvent`.  We don't need to use `waitForCondition` to wait for `SubscriptionState` update now. Thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 4]
```diff
@@ -1443,12 +1443,15 @@ public void assign(Collection<TopicPartition> partitions) {
             // be no following rebalance.
             //
             // See the ApplicationEventProcessor.process() method that handles this event for more detail.
-            applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
-
-            log.info("Assigned to partition(s): {}", partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-            if (subscriptions.assignFromUser(new HashSet<>(partitions)))
-                applicationEventHandler.add(new NewTopicsMetadataUpdateRequestEvent());
+            Timer timer = time.timer(Long.MAX_VALUE);
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Should we used the `defaultApiTimeoutMs` here instead? Just to be sure that in the error case that the assign event does not complete, we don't block forever.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java [hunk 5]
```diff
@@ -16,33 +16,31 @@
...
+public class AssignmentChangeEvent extends CompletableApplicationEvent<Void> {
 
-    private final Map<TopicPartition, OffsetAndMetadata> offsets;
     private final long currentTimeMs;
+    private final Collection<TopicPartition> partitions;
 
-    public AssignmentChangeEvent(final Map<TopicPartition, OffsetAndMetadata> offsets, final long currentTimeMs) {
-        super(Type.ASSIGNMENT_CHANGE);
-        this.offsets = Collections.unmodifiableMap(offsets);
+    public AssignmentChangeEvent(final long currentTimeMs, final long remainingTImeMs, final Collection<TopicPartition> partitions) {
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java]: instead of `remainingTimeMs`, should we pass `deadlineMs` as it's done for the other CompletableEvents? just for consistency

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java [hunk 6]
```diff
@@ -16,33 +16,31 @@
...
 
-import java.util.Collections;
-import java.util.Map;
+import java.util.Collection;
 
-public class AssignmentChangeEvent extends ApplicationEvent {
+public class AssignmentChangeEvent extends CompletableApplicationEvent<Void> {
 
-    private final Map<TopicPartition, OffsetAndMetadata> offsets;
     private final long currentTimeMs;
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java:27]: I notice that this is only needed to `updateAutoCommitTimer` when processing an `AssignmentChangeEvent`. But we already have a `Time`in the `ApplicationEventProcessor`. So, could we simplify here, remove this `currentTimeMs`, and use that time.milliseconds as argument on ln 204 of the AppEventProcessor?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 7]
```diff
@@ -1748,8 +1767,7 @@ public void testMultipleBackgroundErrors() {
         final KafkaException expectedException2 = new KafkaException("Spam, Spam, Spam");
         final ErrorEvent errorEvent2 = new ErrorEvent(expectedException2);
         backgroundEventQueue.add(errorEvent2);
-        consumer.assign(singletonList(new TopicPartition("topic", 0)));
-        final KafkaException exception = assertThrows(KafkaException.class, () -> consumer.poll(Duration.ZERO));
+        final KafkaException exception = assertThrows(KafkaException.class, () -> consumer.assign(singletonList(new TopicPartition("topic", 0))));
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: oh I hadn't realized this undesired side effect of processing background events on assign, but of course, I missed that. I would say we don't want this change in behaviour. 

We only need to ensure that the assign does not return until the background event completes, so not really need to process background events, we just need `applicationEventHandler.addAndGet(assignmentChangeEvent)` (and I would ensure we create the event using the `defaultApiTimeout` for the deadline. Makes sense? we would keep the same behaviour we had on this test (assign does not throw any background event, poll does).

Just for the record, on the unsubscribe we do need to process background events because of the rebalanceCallbacks that run in the app thread, but that's not the case here.

**Review by lianetm — COMMENTED** [review]
Hey @FrankYang0529 , thanks for the updates! Left some comments

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Good catch. I copied this from `unsubscribe`. Do we also want to update it?

https://github.com/apache/kafka/blob/f47b526eb4fe0e4bbf594c08f0e056d7aaaf8e8d/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1477-L1482

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java (hunk 5 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java]: Yes, I replace `remainingTimeMs` with `deadlineMs`. Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java (hunk 6 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java:27]: Sorry, I can't find `Time` object in `ApplicationEventProcessor`, so I still keep `currentTimeMs` here. Do we want to add `Time` object to `ApplicationEventProcessor`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java (hunk 7 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Good catch. Updated it. Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java (hunk 6 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java:27]: oh you're right. I guess I was looking at my local branch when adding that comment (not trunk). I'm adding a `Time` object to appEventProcessor for the changes I'm working on ([KAFKA-17066](https://issues.apache.org/jira/browse/KAFKA-17066)). So let's just leave the currentTime unchanged here, and I'll do the simplification on that other PR that is introducing the Time already.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 8]
```diff
@@ -1416,6 +1415,7 @@ public void assign(Collection<TopicPartition> partitions) {
             if (partitions == null) {
                 throw new IllegalArgumentException("Topic partitions collection to assign to cannot be null");
             }
+            subscriptions.validateAssignFromUserSubscriptionType();
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: do we need this validation here? This is already validated when the actual assignment attempt happens (when we call subscriptionState.assignFromUser in the appEventProcessor). I wonder if we could reuse that, and complete the event exceptionally if that call to assignFromUser fails? 

I expect that then the addAndGet call here on ln 1448 would propagate the exception, so we would achieve the result we want, reusing the validation of the subscriptionState, and using the event as a way to propagate the result (successful or failed). Would that work?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: seems sensible to me, I would expect that no api call should block for more than the defaultApiTimeout unless it takes a timeout as parameter, but I would probably suggest we do that separately, what do you think? Just to have the time to think about it and validate the change (unsubscribe is a lot more complicated, leaved group, executes callbacks, ...)

**Comment by lianetm** [comment]
Hey @FrankYang0529 , thanks for the updates! Left some comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 9]
```diff
@@ -1443,12 +1442,10 @@ public void assign(Collection<TopicPartition> partitions) {
...
-            applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
-
-            log.info("Assigned to partition(s): {}", partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-            if (subscriptions.assignFromUser(new HashSet<>(partitions)))
-                applicationEventHandler.add(new NewTopicsMetadataUpdateRequestEvent());
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            AssignmentChangeEvent assignmentChangeEvent = new AssignmentChangeEvent(timer.currentTimeMs(), calculateDeadlineMs(timer), partitions);
+            applicationEventHandler.addAndGet(assignmentChangeEvent);
+            log.info("Assigned new partitions");
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: nit: should we remove this log line? We already have a very similar one that will show "Assigned to partitions...." when the actual change happens

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java [hunk 10]
```diff
@@ -118,13 +118,12 @@ private static Stream<Arguments> applicationEvents() {
         final long currentTimeMs = 12345;
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java]: this is not used anymore so let's just remove it

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java [hunk 11]
```diff
@@ -118,13 +118,12 @@ private static Stream<Arguments> applicationEvents() {
...
         return Stream.of(
                 Arguments.of(new PollEvent(100)),
-                Arguments.of(new NewTopicsMetadataUpdateRequestEvent()),
                 Arguments.of(new AsyncCommitEvent(new HashMap<>())),
                 Arguments.of(new SyncCommitEvent(new HashMap<>(), 500)),
                 Arguments.of(new ResetPositionsEvent(500)),
                 Arguments.of(new ValidatePositionsEvent(500)),
                 Arguments.of(new TopicMetadataEvent("topic", Long.MAX_VALUE)),
-                Arguments.of(new AssignmentChangeEvent(offset, currentTimeMs)));
+                Arguments.of(new AssignmentChangeEvent(12345, 12345, Collections.emptyList())));
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:131]: note that by adding this we're only testing that the call to `process(AppEvent)` ends up triggering the `process(AssignmentChangeEvent)`, nothing more. So should we add a test for the `process(AssignmentChangeEvent)`? We should ensure that it takes the actions we expect (commit, assign, request metadata update), using a `processor` instance, not a mock.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 12]
```diff
@@ -2182,6 +2207,16 @@ private void completeUnsubscribeApplicationEventSuccessfully() {
         }).when(applicationEventHandler).add(ArgumentMatchers.isA(UnsubscribeEvent.class));
     }
 
+    private void completeAssignmentChangeEventSuccessfully() {
+        doAnswer(invocation -> {
+            AssignmentChangeEvent event = invocation.getArgument(0);
+            // In AsyncKafkaConsumer, the subscription is updated in the background thread.
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: nit: this inline comment doesn't seem to add much imo, the func name and body seem clear enough to me

**Review by lianetm — COMMENTED** [review]
Thanks for the updates @FrankYang0529! A few minor comments and test gap, but really almost there.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java (hunk 11 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:131]: I add `testAssignmentChangeEvent` and `testAssignmentChangeEventWithException` to check `processor` instance. Thank you.

**Comment by FrankYang0529** [comment]
Hi @lianetm, thanks for the review. In previous CI, there is a failed case `PlaintextConsumerCallbackTest#testConsumerRebalanceListenerAssignOnPartitionsAssigned` for consumer group protocol. To keep the behavior, I add some change to `ConsumerUtils#getResult`. Could you take a look when you have time? Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java [hunk 13]
```diff
@@ -231,6 +231,8 @@ public static <T> T getResult(Future<T> future) {
         try {
             return future.get();
         } catch (ExecutionException e) {
+            if (e.getCause() instanceof IllegalStateException)
```
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:236]: interesting, good catch by the integration test. The trick is that this would of course apply to all calls, not only to the assign, but feels like the right thing to ensure we propagate an IllegalState as such, and not obfuscated in a KafkaException.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java [hunk 14]
```diff
@@ -145,6 +148,39 @@ public void testResetPositionsProcess() {
...
+        doReturn(true).when(subscriptionState).assignFromUser(any());
+        processor.process(event);
+        if (withGroupId) {
+            verify(commitRequestManager).updateAutoCommitTimer(currentTimeMs);
+            verify(commitRequestManager).maybeAutoCommitAsync();
+        } else {
+            verify(commitRequestManager, never()).updateAutoCommitTimer(currentTimeMs);
+            verify(commitRequestManager, never()).maybeAutoCommitAsync();
+        }
+        verify(metadata).requestUpdateForNewTopics();
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:169]: let's `verify(subscriptionState).assignFromUser` to ensure the call to update the subscription state actually happens. We could even validate the param because we know it (if we provide a set of partitions to the event in ln 155,  then the assignFromUser call should receive the same set of partitions)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java [hunk 15]
```diff
@@ -145,6 +148,39 @@ public void testResetPositionsProcess() {
...
+        assertDoesNotThrow(() -> event.future().get());
+
```
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java]: nit: remove extra line?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java (hunk 13 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:236]: Do we need to have the same logic here as in the `getResult()` on line 218?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 16]
```diff
@@ -1443,12 +1442,9 @@ public void assign(Collection<TopicPartition> partitions) {
...
             // See the ApplicationEventProcessor.process() method that handles this event for more detail.
-            applicationEventHandler.add(new AssignmentChangeEvent(subscriptions.allConsumed(), time.milliseconds()));
-
-            log.info("Assigned to partition(s): {}", partitions.stream().map(TopicPartition::toString).collect(Collectors.joining(", ")));
-
-            if (subscriptions.assignFromUser(new HashSet<>(partitions)))
-                applicationEventHandler.add(new NewTopicsMetadataUpdateRequestEvent());
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            AssignmentChangeEvent assignmentChangeEvent = new AssignmentChangeEvent(timer.currentTimeMs(), calculateDeadlineMs(timer), partitions);
+            applicationEventHandler.addAndGet(assignmentChangeEvent);
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1447]: With this change we're now blocking on this operation completing. Is that change intentional?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java [hunk 17]
```diff
@@ -145,6 +148,39 @@ public void testResetPositionsProcess() {
...
+    @Test
+    public void testAssignmentChangeEventWithException() {
+        AssignmentChangeEvent event = new AssignmentChangeEvent(12345, 12345, Collections.emptyList());
+
+        setupProcessor(false);
+        doThrow(new IllegalStateException()).when(subscriptionState).assignFromUser(any());
+        processor.process(event);
+
+        ExecutionException e = assertThrows(ExecutionException.class, () -> event.future().get());
+        assertInstanceOf(IllegalStateException.class, e.getCause());
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:183]: Where is the `IllegalStateException` thrown in a production setting? I see we special-cased `IllegalStateException` in `ConsumerUtils`, but I'm wondering if an `IllegalStateException` is something a user can hit 🤔

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the PR @FrankYang0529!

Just a few questions/comments.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 16 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1447]: Yes, this is intentional. In `ClassicKafkaConsumer`, the change for `SubscriptionState` is synchronous. We would like to keep this behavior, so we have to block on it.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java (hunk 17 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:183]: This is in `SubscriptionState`.

https://github.com/apache/kafka/blob/42715654095d019effc478ab343c0132e02aa70f/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java#L230-L231

In `KafkaConsumer#assign` java doc, we also mention this exception.

https://github.com/apache/kafka/blob/42715654095d019effc478ab343c0132e02aa70f/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L783-L784

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java (hunk 14 above)
Comment (FrankYang0529) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java:169]: Yes, updated it. Thank you.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java (hunk 13 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java:236]: I updated it as well. Let's wait for CI result. Thanks for catching it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I create https://issues.apache.org/jira/browse/KAFKA-17377 for followup. Thank you.

**Comment by lianetm** [comment]
I noticed a failure on a relatively related test, `testMissingOffsetNoResetPolicy`, but after digging into it I noticed it has been flaky since it was enabled for the new consumer, before this PR, so I filed [KAFKA-17395](https://issues.apache.org/jira/browse/KAFKA-17395) for it.

**Review by lianetm — APPROVED** [review]
Thanks for the changes @FrankYang0529 , LGTM. Could you please rebase? Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java [hunk 18]
```diff
@@ -16,33 +16,31 @@
...
-        super(Type.ASSIGNMENT_CHANGE);
-        this.offsets = Collections.unmodifiableMap(offsets);
+    public AssignmentChangeEvent(final long currentTimeMs, final long deadlineMs, final Collection<TopicPartition> partitions) {
+        super(Type.ASSIGNMENT_CHANGE, deadlineMs);
         this.currentTimeMs = currentTimeMs;
-    }
-
-    public Map<TopicPartition, OffsetAndMetadata> offsets() {
-        return offsets;
+        this.partitions = partitions;
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java]: Minor point, but we do have a pattern of marking the collections as immutable via `Collections.unmodifiableSet()` and friends.

**Review by kirktrue — APPROVED** [review]
LGTM. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java (hunk 18 above)
Comment (FrankYang0529) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/AssignmentChangeEvent.java]: Thank you. Updated and rebased it.

**Comment by lianetm** [comment]
Test failures unrelated, and 1 failure related to the new consumer (testMissingOffsetNoResetPolicy), but it's been flaky since enabled (graph [here](https://ge.apache.org/scans/tests?search.names=Git%20branch&search.rootProjectNames=kafka&search.startTimeMax=1724990399999&search.startTimeMin=1722484800000&search.timeZoneId=America%2FToronto&search.values=trunk&tests.container=org.apache.kafka.clients.consumer.KafkaConsumerTest&tests.test=testMissingOffsetNoResetPolicy(GroupProtocol)%5B2%5D)) and passes locally, so I'll merge (and file a jira for the flaky as soon as I sort out an issue with my jira account). Thanks @FrankYang0529 !

-- update
Jira for the flaky test https://issues.apache.org/jira/browse/KAFKA-17395
